File: leetcode/2772.使数组中的所有元素都等于零.java
class Solution {
    public boolean checkArray(int[] nums, int k) {
        int n = nums.length;
        int[] f = new int[n+2];
        for(int i = 1; i <= n; i++){
            if(i == 1) f[i] = nums[i-1];
            else f[i] = nums[i-1] - nums[i-2];
        }
        for(int i = 0; i < n-k+1; i++){
            int x = f[i+1] + f[i];
            if(x < 0) return false;
            f[i+1] = 0;
            f[i+1+k] += x;
        }
        for(int i = 1; i <= n; i++){
            if(f[i] != 0) return false;
        }
        return true;
    }
}

File: leetcode/2772.使数组中的所有元素都等于零.py
class Solution:
    def checkArray(self, nums: List[int], k: int) -> bool:
        # 差分数组
        n = len(nums)
        f = [nums[0]] + [0] * n
        for i in range(1, n):
            f[i] = nums[i] - nums[i-1]
        print(f)
        f = [0] + f
        for i in range(n - k + 1):
            x = f[i] + f[i+1]
            if x < 0: return False
            f[i+1] = 0
            f[i+1+k] += x
        print(f)
        for i in range(1, n+1):
            if f[i] != 0 :
                return False
        return True
        